Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift

```diff
@@ -114,3 +330,124 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
+
+    /// Fetches the element in the index before the given one, or an empty array if that would read out of bounds.
+    func getPreviousOrDefault(_ array: [[Element]], index: Int) -> [Element] {
+        if index == 0 {
+            return []
+        } else {
+            return array[index - 1]
+        }
+    }
+
+    /// Returns the longer sequence, or `lhs` if they are of equal length.
+    func maxLength(_ lhs: [Element], _ rhs: [Element]) -> [Element] {
+        if rhs.count > lhs.count {
+            return rhs
+        } else {
+            return lhs
+        }
+    }
+
+    // Iterate row-wise: compare all the elements in the first sequence.
+    for lhsValue in lhs {
+        var currentRow: [[Element]] = .init(repeating: [], count: rhs.count)
+        // Iterate column-wise: with the indices and values from the second sequence, perform the comparison.
+        for (rhsIndex, rhsValue) in rhs.enumerated() {
+            if lhsValue == rhsValue {
+                // If the values are equal, pull the sequence "diagonally up-left" from this coordinate,
+                // and append the current element to it.
+                currentRow[rhsIndex] = getPreviousOrDefault(previousRow, index: rhsIndex) + [lhsValue]
+            } else {
+                // If the values are not equal, pull the longer sequence from either the row above or
+                // the column to the left. Short-circuit by preferring the "row-above" value in case
+                // they're equal.
+                currentRow[rhsIndex] = maxLength(previousRow[rhsIndex], getPreviousOrDefault(currentRow, index: rhsIndex))
+            }
+        }
+        // Now that we've computed this comparison row, save it for the next iteration.
+        previousRow = currentRow
+    }
+
+    return commonPrefix + previousRow.last!
+}
+
+/// Computes the "longest common subsequence" of a list of sequences.
+fileprivate func longestCommonSubsequence<Element>(_ sequences: [[Element]]) -> [Element] where Element: Equatable {
+    var resultSequence: [Element]? = nil
+
+    // This function relies on the two-sequence LCS to expand to an arbitrary set of sequences. The
+    // idea is effectively a `reduce1` without a default case. This does expand the worst-case time
+    // complexity to nearly `L^N` comparisons, with `L` being the average length of the sequences,
+    // and `N` being the number of sequences. In practice, since overloaded declarations tend to
+    // share a common prefix (which is trimmed in the underlying implementation), and because this
+    // is a rolling comparison against the calculated LCS, the actual number of comparisons is
+    // likely to be less than that in the average case.
+    for sequence in sequences {
+        if let currentResult = resultSequence {
+            resultSequence = longestCommonSubsequence(currentResult, sequence)
+        } else {
+            resultSequence = sequence
+        }
+    }
+    return resultSequence ?? []
+}
```

Comment:
I haven't had time to profile this yet, but there's built-in diffing support in Swift which can be used to compute the longest common subsequence in a few lines by incrementally applying only the removals to the first value:

```swift
func longestCommonSubsequence<Element: Equatable>(_ sequences: [[Element]]) -> [Element] {
    guard var result = sequences.first else { return [] }
    
    for other in sequences.dropFirst() {
        for case .remove(let index, _, _) in other.difference(from: result).removals {
            result.remove(at: index)
        }
    }
    return result
}
```

It's possible that there's some overhead from computing insertions and never using them but [the underlying implementation](https://github.com/apple/swift/blob/main/stdlib/public/core/Diffing.swift#L228) also looks well optimized.

Reply:
It looks like this implementation is faster than my naive version. 😅 I copy-pasted your version into the code in place of my *n*-sequence wrapper and ran a benchmark between this and my version, running on the same "large framework with many overloads". It looks like there's a consistent speed-up:

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ bench-custom         │ bench-stdlib         │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ no change¹,²    │ 8.219 sec            │ 8.167 sec            │
│ Duration for 'convert-total-time'        │ -1.703%³,⁴      │ 12.8 sec             │ 12.582 sec           │
│ Duration for 'documentation-processing'  │ -3.751%⁵        │ 4.471 sec            │ 4.303 sec            │
│ Duration for 'finalize-navigation-index' │ no change⁶      │ 0.038 sec            │ 0.041 sec            │
│ Peak memory footprint                    │ no change⁷,⁸    │ 717.4 MB             │ 726.3 MB             │
│ Data subdirectory size                   │ +0.003461%⁹     │ 130 MB               │ 130 MB               │
│ Index subdirectory size                  │ no change¹⁰     │ 1.1 MB               │ 1.1 MB               │
│ Total DocC archive size                  │ +0.00209%¹¹     │ 232.4 MB             │ 232.4 MB             │
│ Topic Anchor Checksum                    │ no change       │ 53074b40863239b7a00d │ 53074b40863239b7a00d │
│ Topic Graph Checksum                     │ no change       │ 5bdfe323b77baec3d5cd │ 5bdfe323b77baec3d5cd │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Here's the new comparison between this PR and `main`:

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ main                 │ current              │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ +1.913%¹,²      │ 10.351 sec           │ 10.549 sec           │
│ Duration for 'convert-total-time'        │ +3.857%³,⁴      │ 15.477 sec           │ 16.074 sec           │
│ Duration for 'documentation-processing'  │ +7.959%⁵,⁶      │ 5.003 sec            │ 5.401 sec            │
│ Duration for 'finalize-navigation-index' │ no change⁷      │ 0.046 sec            │ 0.046 sec            │
│ Peak memory footprint                    │ no change⁸      │ 811 MB               │ 835.4 MB             │
│ Data subdirectory size                   │ +2.028%⁹        │ 127.4 MB             │ 130 MB               │
│ Index subdirectory size                  │ -0.02829%¹⁰     │ 1.1 MB               │ 1.1 MB               │
│ Total DocC archive size                  │ +1.092%¹¹       │ 236.6 MB             │ 239.2 MB             │
│ Topic Anchor Checksum                    │ no change       │ 53074b40863239b7a00d │ 53074b40863239b7a00d │
│ Topic Graph Checksum                     │ change          │ 5bdfe323b77baec3d5cd │ f4af046ddef9eab49ba8 │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```